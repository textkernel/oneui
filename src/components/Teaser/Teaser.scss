@import '../../themes/oneui/oneui';

.Teaser {
    min-width: 0;
    margin: 0;
    display: flex;
    flex: 1;
    flex-direction: column;
    height: 63px;
    color: var(--color-neutral-60);

    &__line {
        display: flex;
        flex: 1;
        align-items: center;
        padding-bottom: var(--spacing-2x);

        &:last-child {
            padding-bottom: 0;
            justify-content: flex-end;
        }
    }

    &__title {
        font-weight: var(--font-weight-bold);
        white-space: nowrap;
        overflow: hidden;
        text-overflow: ellipsis;
    }
    &__location {
        white-space: nowrap;

        &::before {
            content: '\00a0\2014\00a0';
        }
    }

    &__subTitle {
        white-space: nowrap;
        overflow: hidden;
        text-overflow: ellipsis;
    }

    &__details {
        flex: 1;
        white-space: nowrap;
        overflow: hidden;
        text-overflow: ellipsis;
    }

    &__statuses {
        padding-left: var(--spacing-5x);
    }
}
