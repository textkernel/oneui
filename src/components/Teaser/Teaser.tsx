import * as React from 'react';
import { bem } from '../../utils';
import { Text } from '../Text';
import { Tooltip } from '../Tooltip';
import styles from './Teaser.scss';

type Status = {
    label: string;
    tooltip: string;
};

export interface Props extends Omit<React.HTMLAttributes<HTMLDivElement>, 'title'> {
    /** The title of entity. Note: it is rendered within a <span> */
    title: React.ReactNode;
    /** The subtitle of the entity. Note: it is rendered within a <span> */
    subTitle?: React.ReactNode;
    /** The location of the entity. Note: it is rendered within a <span> */
    location?: React.ReactNode;
    /** Details related to the entity. Note: it is rendered within a <span> */
    details?: React.ReactNode;
    /** Statuses of entity containing status label and tooltip message */
    statuses?: Status[];
    /** Style the component as disabled */
    disabled?: boolean;
}

const { block, elem } = bem('Teaser', styles);

export const Teaser: React.FC<Props> = ({
    title,
    subTitle = '',
    location = '',
    details = '',
    statuses,
    disabled = false,
    ...rest
}) => {
    const subTitleElement = (
        <Text
            inline
            title={typeof subTitle === 'string' ? subTitle : undefined}
            context={disabled ? 'neutral' : 'background'}
            {...elem('subTitle')}
        >
            {subTitle}
        </Text>
    );
    const detailsElement = (
        <Text
            inline
            title={typeof details === 'string' ? details : undefined}
            context="neutral"
            {...elem('details')}
        >
            {details}
        </Text>
    );
    return (
        <div {...rest} {...block({ ...rest })}>
            <div {...elem('line')}>
                <Text
                    inline
                    context={disabled ? 'neutral' : 'foreground'}
                    {...elem('title')}
                    title={typeof title === 'string' ? title : undefined}
                >
                    {title}
                </Text>
                {location && (
                    <Text inline context="neutral" {...elem('location')}>
                        {location}
                    </Text>
                )}
            </div>
            <div {...elem('line')}>
                {subTitle && subTitleElement}
                {!subTitle && details && detailsElement}
            </div>
            <div {...elem('line')}>
                {subTitle && details && detailsElement}
                {statuses && statuses.length > 0 && (
                    <div {...elem('statuses')}>
                        {statuses.map(({ label, tooltip }, index) => (
                            <span key={label}>
                                <Tooltip content={tooltip}>
                                    <Text inline context="neutral" size="small">
                                        {label}
                                    </Text>
                                </Tooltip>
                                {index < statuses.length - 1 && (
                                    <Text inline context="neutral" size="small">
                                        {' & '}
                                    </Text>
                                )}
                            </span>
                        ))}
                    </div>
                )}
            </div>
        </div>
    );
};

Teaser.displayName = 'Teaser';
