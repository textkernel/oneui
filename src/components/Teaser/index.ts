export { Teaser, Props as TeaserProps } from './Teaser';
