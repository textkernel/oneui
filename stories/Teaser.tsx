import * as React from 'react';
import { storiesOf } from '@storybook/react';
import { boolean, object, text, withKnobs } from '@storybook/addon-knobs';
import { Teaser } from '@textkernel/oneui';

storiesOf('Molecules|Teaser', module)
    .addDecorator(withKnobs)
    .add('With all fields', () => {
        const title = text('Title', 'My first job');
        return (
            <Teaser
                title={<span title={title}>{title}</span>}
                subTitle={text('SubTitle', 'Awsome inc.')}
                location={text('Location', 'Melbourne')}
                details={text('Details', 'It was posted here')}
                statuses={[
                    object('Viewed status', {
                        label: 'Viewed',
                        tooltip: 'Viewed one day ago',
                    }),
                    object('Imported status', {
                        label: 'Imported',
                        tooltip: 'Imported two days ago',
                    }),
                ]}
                disabled={boolean('Disabled', false)}
            />
        );
    })
    .add('With one status', () => {
        const title = text('Title', 'My first job');
        return (
            <Teaser
                title={<span title={title}>{title}</span>}
                subTitle={text('SubTitle', 'Awsome inc.')}
                location={text('Location', 'Melbourne')}
                details={text('Details', 'It was posted here')}
                statuses={[
                    object('Viewed status', {
                        label: 'Viewed',
                        tooltip: 'Viewed one day ago',
                    }),
                ]}
                disabled={boolean('Disabled', false)}
            />
        );
    })
    .add('Without statuses', () => {
        const title = text('Title', 'My first job');
        return (
            <Teaser
                title={<span title={title}>{title}</span>}
                subTitle={text('SubTitle', 'Awsome inc.')}
                location={text('Location', 'Melbourne')}
                details={text('Details', 'It was posted here')}
                disabled={boolean('Disabled', false)}
            />
        );
    });
