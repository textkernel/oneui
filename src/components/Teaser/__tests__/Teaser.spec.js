import React from 'react';
import toJson from 'enzyme-to-json';
import { Teaser } from '../Teaser';

describe('Teaser', () => {
    it('should render correctly with just a title', () => {
        const wrapper = mount(<Teaser title="some title" />);

        expect(toJson(wrapper)).toMatchSnapshot();
    });
    it('should render with all props defined', () => {
        const wrapper = mount(
            <Teaser
                title="A job title"
                location="location"
                subTitle="Organization"
                details="details about this job"
                statuses={[
                    {
                        label: 'Viewed',
                        tooltip: 'Viewed two days ago',
                    },
                    {
                        label: 'Imported',
                        tooltip: 'Imported five days ago',
                    },
                ]}
            />
        );

        expect(toJson(wrapper)).toMatchSnapshot();
        expect(wrapper.find('Text').at(0).prop('context')).toEqual('brand');
        expect(wrapper.find('Text').at(2).prop('context')).toEqual('accent');
        expect(wrapper.find('Text').at(0).prop('title')).toEqual('A job title');
    });
    it('should render correctly in disabled mode', () => {
        const wrapper = mount(
            <Teaser
                title="A job title"
                location="location"
                subTitle="Organization"
                details="details about this job"
                disabled
            />
        );

        expect(toJson(wrapper)).toMatchSnapshot();
        expect(wrapper.find('Text').at(0).prop('context')).toEqual('muted');
        expect(wrapper.find('Text').at(2).prop('context')).toEqual('muted');
    });
});
