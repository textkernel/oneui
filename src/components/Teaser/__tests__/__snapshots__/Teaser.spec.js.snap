// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Teaser should render correctly in disabled mode 1`] = `
<Teaser
  details="details about this job"
  disabled={true}
  location="location"
  subTitle="Organization"
  title="A job title"
>
  <div
    className="Teaser"
  >
    <div
      className="Teaser__line"
    >
      <Text
        className="Teaser__title"
        context="muted"
        inline={true}
        title="A job title"
      >
        <span
          className="Text--context_muted Teaser__title"
          title="A job title"
        >
          A job title
        </span>
      </Text>
      <Text
        className="Teaser__location"
        context="muted"
        inline={true}
      >
        <span
          className="Text--context_muted Teaser__location"
        >
          location
        </span>
      </Text>
    </div>
    <div
      className="Teaser__line"
    >
      <Text
        className="Teaser__subTitle"
        context="muted"
        inline={true}
        title="Organization"
      >
        <span
          className="Text--context_muted Teaser__subTitle"
          title="Organization"
        >
          Organization
        </span>
      </Text>
    </div>
    <div
      className="Teaser__line"
    >
      <Text
        className="Teaser__details"
        context="muted"
        inline={true}
        title="details about this job"
      >
        <span
          className="Text--context_muted Teaser__details"
          title="details about this job"
        >
          details about this job
        </span>
      </Text>
    </div>
  </div>
</Teaser>
`;

exports[`Teaser should render correctly with just a title 1`] = `
<Teaser
  title="some title"
>
  <div
    className="Teaser"
  >
    <div
      className="Teaser__line"
    >
      <Text
        className="Teaser__title"
        context="brand"
        inline={true}
        title="some title"
      >
        <span
          className="Text--context_brand Teaser__title"
          title="some title"
        >
          some title
        </span>
      </Text>
    </div>
    <div
      className="Teaser__line"
    />
    <div
      className="Teaser__line"
    />
  </div>
</Teaser>
`;

exports[`Teaser should render with all props defined 1`] = `
<Teaser
  details="details about this job"
  location="location"
  statuses={
    Array [
      Object {
        "label": "Viewed",
        "tooltip": "Viewed two days ago",
      },
      Object {
        "label": "Imported",
        "tooltip": "Imported five days ago",
      },
    ]
  }
  subTitle="Organization"
  title="A job title"
>
  <div
    className="Teaser"
  >
    <div
      className="Teaser__line"
    >
      <Text
        className="Teaser__title"
        context="brand"
        inline={true}
        title="A job title"
      >
        <span
          className="Text--context_brand Teaser__title"
          title="A job title"
        >
          A job title
        </span>
      </Text>
      <Text
        className="Teaser__location"
        context="muted"
        inline={true}
      >
        <span
          className="Text--context_muted Teaser__location"
        >
          location
        </span>
      </Text>
    </div>
    <div
      className="Teaser__line"
    >
      <Text
        className="Teaser__subTitle"
        context="accent"
        inline={true}
        title="Organization"
      >
        <span
          className="Text--context_accent Teaser__subTitle"
          title="Organization"
        >
          Organization
        </span>
      </Text>
    </div>
    <div
      className="Teaser__line"
    >
      <Text
        className="Teaser__details"
        context="muted"
        inline={true}
        title="details about this job"
      >
        <span
          className="Text--context_muted Teaser__details"
          title="details about this job"
        >
          details about this job
        </span>
      </Text>
      <div
        className="Teaser__statuses"
      >
        <span
          key="Viewed"
        >
          <Tooltip
            animation="shift-toward"
            content="Viewed two days ago"
          >
            <ForwardRef(TippyWrapper)
              animation="shift-toward"
              content="Viewed two days ago"
            >
              <Tippy
                animation="shift-toward"
                content="Viewed two days ago"
              >
                <Text
                  context="muted"
                  inline={true}
                  size="small"
                >
                  <span
                    className="Text--context_muted Text--size_small"
                  >
                    Viewed
                  </span>
                </Text>
                <Portal
                  containerInfo={
                    <div>
                      Viewed two days ago
                    </div>
                  }
                >
                  Viewed two days ago
                </Portal>
              </Tippy>
            </ForwardRef(TippyWrapper)>
          </Tooltip>
          <Text
            context="muted"
            inline={true}
            size="small"
          >
            <span
              className="Text--context_muted Text--size_small"
            >
               & 
            </span>
          </Text>
        </span>
        <span
          key="Imported"
        >
          <Tooltip
            animation="shift-toward"
            content="Imported five days ago"
          >
            <ForwardRef(TippyWrapper)
              animation="shift-toward"
              content="Imported five days ago"
            >
              <Tippy
                animation="shift-toward"
                content="Imported five days ago"
              >
                <Text
                  context="muted"
                  inline={true}
                  size="small"
                >
                  <span
                    className="Text--context_muted Text--size_small"
                  >
                    Imported
                  </span>
                </Text>
                <Portal
                  containerInfo={
                    <div>
                      Imported five days ago
                    </div>
                  }
                >
                  Imported five days ago
                </Portal>
              </Tippy>
            </ForwardRef(TippyWrapper)>
          </Tooltip>
        </span>
      </div>
    </div>
  </div>
</Teaser>
`;
